subsys/net/l2/ieee802154/ieee802154.c

```diff
@@ -119,6 +119,75 @@ static inline void set_pkt_ll_addr(struct net_linkaddr *addr, bool comp,
 	addr->type = NET_LINK_IEEE802154;
 }
 
+/**
+ * Filters the destination address of the frame (used when IEEE802154_HW_FILTER
+ * is not available).
+ */
+static bool ieeee802154_check_dst_addr(struct net_if *iface,
+				      struct ieee802154_mhr *mhr)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+	const struct ieee802154_radio_api *radio =
+		net_if_get_device(iface)->api;
+	uint8_t i;
+
+	/*
+	 * FIXME: should I remove this check so the destination address is
+	 * always filtered? */
+	if (radio && (radio->get_capabilities(net_if_get_device(iface)) &
+		      IEEE802154_HW_FILTER)) {
+		return true;
+	}
+
+	/*
+	 * Apply filtering requirements from chapter 6.7.2 of the 802.15.4 standard:
+	 */
+
+	/*
+	 * d.3. The Destination Address field and the Destination PAN ID field are
+	 * not included in the frame => macImplicitBroadcast is TRUE
+	 */
+	if (mhr->fs->fc.dst_addr_mode == IEEE802154_ADDR_MODE_NONE) {
+		return true;
+	}
+
+	/*
+	 * c. If a destination PAN ID is included in the frame, it shall match
+	 * macPanId or shall be the broadcastPAN ID
+	 */
+	if (!(mhr->dst_addr->plain.pan_id == 0xFFFF
+			|| mhr->dst_addr->plain.pan_id == ctx->pan_id)) {
+		LOG_WRN("Frame PAN ID does not match!");
+		return false;
+		}
+
+	if (mhr->fs->fc.dst_addr_mode == IEEE802154_ADDR_MODE_SHORT) {
+		/*
+		 * d.1. A short destination address is included in the frame, and it
+		 * matches either macShortAddress orthe broadcast address.
+		 */
+		if (mhr->dst_addr->plain.addr.short_addr) {
```

Comment:
It looks something is missing here. This condition is not aligned with d.1 description above.